### PR TITLE
`azurerm_windows_web_app`, `azurerm_windows_web_app_slot`, `azurerm_linux_web_app`, `azurerm_linux_web_app_slot` - remove `auto_heal_enabled` property

### LINKED ISSUE #21643: auto_heal_settings should not be required if auto_heal_enabled is false
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

1.4.6

### AzureRM Provider Version

3.5.4

### Affected Resource(s)/Data Source(s)

azurerm_linux_web_app, and likely other app service related resources

### Terraform Configuration Files

```hcl
resource "azurerm_linux_web_app" "web_app" {
  <minimal configuration of web_app>
  auto_heal_enabled = false
}
```

### Debug Output/Panic Output

```shell
│ "site_config.0.auto_heal_enabled": all of
│ `site_config.0.auto_heal_enabled,site_config.0.auto_heal_setting` must be
│ specified
```

### Expected Behaviour

If `auto_heal_enabled` is explicitly false, behavior should be the same as if it is not set.
Resource should only require a auto_heal_settings block if the value is explicitly true.

### Actual Behaviour

Resource fails to deploy with error about missing auto heal settings even though auto_heal_enabled is explicitly false.

### Steps to Reproduce

_No response_

### Important Factoids

_No response_

### References

_No response_

### PR DESCRIPTION
**Pull request by xiaxyi**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review

## Description

Currently, the `auto_heal_enabled` is required to be used together with the `auto_heal_setting` block. Even the user sets the `auto_heal_enabled` to `false`, the provider still ask the user to input the `auto_heal_setting` which is not very reasonable.

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [x] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

image
## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* `azurerm_resource` - support for the `thing1` property [GH-00000]

This is a (please select all that apply):

- [x] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [ ] Enhancement
- [x] Breaking Change

## Related Issue(s)
Fixes #21643 

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

File: internal/services/appservice/migration/linux_web_app.go [hunk 1]
```diff
@@ -1249,7 +1244,7 @@ func (l LinuxWebAppV0toV1) Schema() map[string]*pluginsdk.Schema {
 											},
 
 											"status_code": {
-												Type:     pluginsdk.TypeList,
+												Type:     pluginsdk.TypeSet,
```
Comment (jackofallops) [inline internal/services/appservice/migration/linux_web_app.go:1247]: This needs to be put back to `TypeList` as it represents the schema at the point in time of the migration, where it was a list.

File: internal/services/appservice/migration/linux_web_app_slot.go [hunk 2]
```diff
@@ -1249,7 +1244,7 @@ func (l LinuxWebAppSlotV0toV1) Schema() map[string]*pluginsdk.Schema {
 											},
 
 											"status_code": {
-												Type:     pluginsdk.TypeList,
+												Type:     pluginsdk.TypeSet,
```
Comment (jackofallops) [inline internal/services/appservice/migration/linux_web_app_slot.go:1247]: as above.

**Review by jackofallops — CHANGES_REQUESTED** [review]
HI @xiaxyi - Not a full review, just noticed in passing you seem to have touched the state migrations which will need to be reverted as noted below.  Someone will loop back for a full review as soon as they're able.

thanks

File: internal/services/appservice/helpers/linux_web_app_schema.go [hunk 3]
```diff
@@ -1174,6 +1158,10 @@ func (s *SiteConfigLinux) ExpandForUpdate(metadata sdk.ResourceMetaData, existin
 	}
 
 	if metadata.ResourceData.HasChange("site_config.0.auto_heal_setting") {
+		expanded.AutoHealEnabled = pointer.To(false)
```
Comment (jackofallops) [inline internal/services/appservice/helpers/linux_web_app_schema.go:1161]: This is already `false` from the top of the function, can we check the logic here and remove one of these?

File: internal/services/appservice/helpers/web_app_slot_schema.go [hunk 4]
```diff
@@ -867,6 +849,10 @@ func (s *SiteConfigLinuxWebAppSlot) ExpandForUpdate(metadata sdk.ResourceMetaDat
 	}
 
 	if metadata.ResourceData.HasChange("site_config.0.auto_heal_setting") {
+		expanded.AutoHealEnabled = pointer.To(false)
```
Comment (jackofallops) [inline internal/services/appservice/helpers/web_app_slot_schema.go:852]: as above, if there's no change we're automatically turning auto-heal off, so we should only set it here, not at the top of the function?

File: internal/services/appservice/helpers/windows_web_app_schema.go [hunk 5]
```diff
@@ -828,7 +813,12 @@ func (s *SiteConfigWindows) ExpandForUpdate(metadata sdk.ResourceMetaData, exist
 		}
 		expanded.Cors = cors
 	}
+
 	if metadata.ResourceData.HasChange("site_config.0.auto_heal_setting") {
+		expanded.AutoHealEnabled = pointer.To(false)
```
Comment (jackofallops) [inline internal/services/appservice/helpers/windows_web_app_schema.go:818]: again here, we should double check the logic, but I believe we'll need to not set `false` at the top of the function or we'll possibly switch off auto-heal unexpectedly?

**Review by jackofallops — CHANGES_REQUESTED** [review]
Thanks @xiaxyi for the changes. I've spotted another potential problem below if you can take a look?

Also, since this is a breaking change, we'll target merge after the final 3.x release ships in a couple weeks so this will only be available in 4.0.

File: internal/services/appservice/helpers/windows_web_app_schema.go (hunk 5 above)
Comment (xiaxyi) [reply internal/services/appservice/helpers/windows_web_app_schema.go:818]: Thanks Steve, my first thought is to check the situation like:
1. the `auto_heal` is enabled
2. there is a change to another property under `site_config`

We can't turn the auto heal off under the above mentioned situation during the app update, but change the auto heal state only when the change is targeted to `auto_heal` setting.

Does it makes sense to you?

**Comment by xiaxyi** [comment]
Thanks @jackofallops for the review! I left my comment  in the thread. Let me know if there is anything unclear to you.

**Review by jackofallops — APPROVED** [review]
Thanks @xiaxyi - I think this is fine now. Since it's a breaking change, it won't make it into the last 3.x release, so we'll look to merge this as part of the 4.0 release expected next week.

**Comment by jackofallops** [comment]
Tests looking fine, failures are service capacity related:

image

**Pull request merged**